File: backend/app/schemas/ml_models.py
"""Phase 3 — #15: Pydantic v2 schemas for the ML Model Registry API."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class ModelType(str, Enum):
    FRAUD_CLASSIFIER = "fraud_classifier"
    ANOMALY_DETECTOR = "anomaly_detector"
    RISK_SCORER = "risk_scorer"
    BEHAVIORAL_PROFILER = "behavioral_profiler"


class ModelStatus(str, Enum):
    TRAINING = "training"
    VALIDATING = "validating"
    ACTIVE = "active"
    RETIRED = "retired"
    FAILED = "failed"


class ModelMetrics(BaseModel):
    accuracy: Optional[float] = Field(None, ge=0.0, le=1.0)
    precision: Optional[float] = Field(None, ge=0.0, le=1.0)
    recall: Optional[float] = Field(None, ge=0.0, le=1.0)
    f1: Optional[float] = Field(None, ge=0.0, le=1.0)
    auc_roc: Optional[float] = Field(None, ge=0.0, le=1.0)
    auc_pr: Optional[float] = Field(None, ge=0.0, le=1.0)
    n_train_samples: Optional[int] = None

    model_config = {"extra": "allow"}   # allow additional metrics


class RegisteredModelResponse(BaseModel):
    id: str
    name: str
    model_type: ModelType
    version: str
    status: ModelStatus
    metrics: Dict = Field(default_factory=dict)
    artifact_path: str = ""
    training_dataset_info: Dict = Field(default_factory=dict)
    description: str = ""
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RegisterModelRequest(BaseModel):
    name: str = Field(..., min_length=1)
    model_type: ModelType
    version: str = Field(..., min_length=1)
    description: str = ""
    metrics: Dict = Field(default_factory=dict)
    artifact_path: str = ""
    training_dataset_info: Dict = Field(default_factory=dict)


class UpdateMetricsRequest(BaseModel):
    metrics: Dict = Field(..., description="Metrics to merge into existing record")


class RetrainRequest(BaseModel):
    model_type: ModelType
    n_samples: int = Field(10_000, ge=1_000, le=200_000)
    reason: str = Field("", description="Why retraining was triggered")


class RetrainResponse(BaseModel):
    job_id: str
    model_type: ModelType
    status: str
    message: str
    metrics: Dict = Field(default_factory=dict)


class InferenceRequest(BaseModel):
    features: List[List[float]] = Field(
        ..., description="2-D array: [[f1, f2, ...], ...]"
    )


class InferenceResponse(BaseModel):
    model_id: str
    model_name: str
    version: str
    probabilities: List[float]
    predictions: List[bool]
    latency_ms: float
    backend: str

File: frontend/src/types/ml-models.ts
// Phase 3 — ML Model Registry: TypeScript types (F2.7 / F2.8 / F2.9)

export type ModelType =
  | "fraud_classifier"
  | "anomaly_detector"
  | "risk_scorer"
  | "behavioral_profiler";

export type ModelStatus =
  | "training"
  | "validating"
  | "active"
  | "retired"
  | "failed";

export interface ModelMetrics {
  accuracy?: number;
  precision?: number;
  recall?: number;
  f1?: number;
  auc_roc?: number;
  auc_pr?: number;
  n_train_samples?: number;
  [key: string]: number | undefined;
}

export interface RegisteredModel {
  id: string;
  name: string;
  model_type: ModelType;
  version: string;
  status: ModelStatus;
  metrics: ModelMetrics;
  artifact_path: string;
  training_dataset_info: Record<string, unknown>;
  description: string;
  created_at: string;
  updated_at: string;
}

export interface RegisterModelRequest {
  name: string;
  model_type: ModelType;
  version: string;
  description?: string;
  metrics?: ModelMetrics;
  artifact_path?: string;
  training_dataset_info?: Record<string, unknown>;
}

export interface RetrainRequest {
  model_type: ModelType;
  n_samples?: number;
  reason?: string;
}

export interface RetrainResponse {
  job_id: string;
  model_type: ModelType;
  status: string;
  message: string;
  metrics: ModelMetrics;
}

export interface InferenceRequest {
  features: number[][];
}

export interface InferenceResponse {
  model_id: string;
  model_name: string;
  version: string;
  probabilities: number[];
  predictions: boolean[];
  latency_ms: number;
  backend: string;
}

// UI helpers
export const MODEL_TYPE_LABELS: Record<ModelType, string> = {
  fraud_classifier: "Fraud Classifier",
  anomaly_detector: "Anomaly Detector",
  risk_scorer: "Risk Scorer",
  behavioral_profiler: "Behavioral Profiler",
};

export const MODEL_STATUS_CONFIG: Record<
  ModelStatus,
  { label: string; color: string; bg: string }
> = {
  active: { label: "Active", color: "text-green-700", bg: "bg-green-100" },
  validating: { label: "Validating", color: "text-blue-700", bg: "bg-blue-100" },
  training: { label: "Training", color: "text-yellow-700", bg: "bg-yellow-100" },
  retired: { label: "Retired", color: "text-gray-600", bg: "bg-gray-100" },
  failed: { label: "Failed", color: "text-red-700", bg: "bg-red-100" },
};

export const KEY_METRICS: Array<{ key: keyof ModelMetrics; label: string }> = [
  { key: "accuracy", label: "Accuracy" },
  { key: "precision", label: "Precision" },
  { key: "recall", label: "Recall" },
  { key: "f1", label: "F1" },
  { key: "auc_roc", label: "AUC-ROC" },
  { key: "auc_pr", label: "AUC-PR" },
];
